src/coreclr/jit/fginline.cpp

```diff
@@ -586,8 +586,70 @@ class SubstitutePlaceholdersAndDevirtualizeWalker : public GenTreeVisitor<Substi
                 }
 
                 CORINFO_CONTEXT_HANDLE contextInput = context;
+                context                             = nullptr;
                 m_compiler->impDevirtualizeCall(call, nullptr, &method, &methodFlags, &contextInput, &context,
                                                 isLateDevirtualization, explicitTailCall);
+                if (context != nullptr && !m_compiler->compCurBB->HasFlag(BBF_INTERNAL))
+                {
+                    CORINFO_METHOD_INFO methInfo;
+                    if (m_compiler->info.compCompHnd->getMethodInfo(call->gtCallMethHnd, &methInfo, context) &&
+                        genActualType(JITtype2varType(methInfo.args.retType)) == genActualType(call->TypeGet()))
+                    {
+                        IL_OFFSET ilOffset = 0;
+                        INDEBUG(ilOffset = call->gtRawILOffset);
+                        CORINFO_CALL_INFO callInfo = {};
+                        callInfo.hMethod           = method;
+                        callInfo.methodFlags       = methodFlags;
+                        m_compiler->impMarkInlineCandidate(call, context, false, &callInfo, ilOffset);
+
+                        const bool isInlineCandidate                  = call->IsInlineCandidate();
+                        const bool isGuardedDevirtualizationCandidate = call->IsGuardedDevirtualizationCandidate();
+
+                        if (isInlineCandidate || isGuardedDevirtualizationCandidate)
+                        {
+                            if (parent != nullptr || genActualType(call->TypeGet()) != TYP_VOID)
+                            {
+                                Statement* stmt = m_compiler->gtNewStmt(call);
+                                m_compiler->fgInsertStmtBefore(m_compiler->compCurBB, m_compiler->compCurStmt, stmt);
+                                GenTreeRetExpr* retExpr =
+                                    m_compiler->gtNewInlineCandidateReturnExpr(call->AsCall(),
+                                                                               genActualType(call->TypeGet()));
+
+                                if (call->IsGuardedDevirtualizationCandidate())
+                                {
+                                    for (uint8_t i = 0; i < call->GetInlineCandidatesCount(); i++)
+                                    {
+                                        call->GetGDVCandidateInfo(i)->retExpr = retExpr;
+                                    }
+                                }
+                                else
+                                {
+                                    call->GetSingleInlineCandidateInfo()->retExpr = retExpr;
+                                }
+
+                                m_compiler->compCurStmt = stmt;
```

Comment:
Instead of promoting `comCurStmt` we can have the walker record this statement and handle backing up that way; that seems a bit cleaner to me.

Eg in `fgInline`
```
 walker.WalkTree(...);
 if (walker.AddedInlineCandidates())
 {
     stmt = walker.GetStmt();
 }
```